src/server/tiering/small_bins.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/small_bins.h"
+
+#include <algorithm>
+#include <optional>
+#include <utility>
+
+#include "absl/base/internal/endian.h"
+#include "base/logging.h"
+#include "core/compact_object.h"
+#include "server/tiering/disk_storage.h"
+
+namespace dfly::tiering {
+
+std::optional<SmallBins::FilledBin> SmallBins::Stash(std::string_view key, std::string_view value) {
+  DCHECK_LT(value.size(), 2_KB);
+
+  size_t value_bytes = 8 /* hash */ + 2 /* length */ + value.size();
+
+  std::optional<FilledBin> filled_bin;
+  if (current_bin_bytes_ + value_bytes >= 4_KB) {
+    filled_bin = FlushBin();
+  }
+
+  current_bin_bytes_ += value_bytes;
+  current_bin_[key] = value;
+  return filled_bin;
+}
+
+SmallBins::FilledBin SmallBins::FlushBin() {
+  std::string out;
+  out.resize(current_bin_bytes_ + 1);
+
+  unsigned id = ++last_bin_id_;
+  auto& pending_list = pending_bins_[id];
+
+  char* data = out.data();
+  while (!current_bin_.empty()) {
+    auto node = current_bin_.extract(current_bin_.begin());
+    auto& key = node.key();
+    auto& value = node.mapped();
+
+    absl::little_endian::Store64(data, CompactObj::HashCode(key));
+    data += sizeof(uint64_t);
+
+    absl::little_endian::Store16(data, static_cast<uint16_t>(value.size()));
+    data += sizeof(uint16_t);
+
+    memcpy(data, value.data(), value.size());
+    data += value.size();
+
+    pending_list.emplace_back(key, value.size());
+  }
+
+  current_bin_bytes_ = 0;
+  current_bin_.clear();
+
+  return {id, std::move(out)};
+}
+
+SmallBins::CutList SmallBins::ReportStashed(unsigned id, DiskSegment segment) {
```

Comment:
what's a cutlist? why this name?

Reply:
I'll call it KeySegments. After a bin was stashed, this list returns where each value is stored on the page (segment)